### ISSUE TITLE
scylla_sysconfig_setup => scylla_util.py: perftune_base_command: at least one directory has to be set in 'data_file_directory'

### ISSUE THREAD

**Issue by amoskong** [scylladb/scylla#5357]
*Installation details*
Scylla version (or git commit hash): scylla-666.development-0.20191126.4160b3630d.el7.x86_64
Cluster size: 1
OS (RHEL/CentOS/Ubuntu/AWS AMI): ami-0bcf61ab6b989903d (us-east-1)

```
$ rpm -qa |grep scylla
scylla-python3-3.7.5-0.20191127.4160b3630d.el7.x86_64
scylla-server-666.development-0.20191126.4160b3630d.el7.x86_64
scylla-666.development-0.20191126.4160b3630d.el7.x86_64
scylla-jmx-666.development-20191127.d8c47603d9.noarch
scylla-conf-666.development-0.20191126.4160b3630d.el7.x86_64
scylla-kernel-conf-666.development-0.20191126.4160b3630d.el7.x86_64
scylla-tools-666.development-20191127.35906df2a0.noarch
scylla-debuginfo-666.development-0.20191126.4160b3630d.el7.x86_64
scylla-tools-core-666.development-20191127.35906df2a0.noarch
scylla-ami-666.development-20191127.15bb0a7.fc27.noarch
```

- AMI test job: https://jenkins.scylladb.com/job/scylla-master/job/ami-test/77/
- AMI test log: https://jenkins.scylladb.com/job/scylla-master/job/ami-test/77/artifact/latest/job.log

### Description:

scylla ami setup failed, scylla-server isn't started, test failed in prepare stage.

```
-- Logs begin at Wed 2019-11-27 07:45:44 UTC, end at Wed 2019-11-27 08:10:55 UTC. --
Nov 27 07:46:56 ip-172-16-99-162.ec2.internal systemd[1]: Starting Scylla AMI Setup...
Nov 27 07:47:10 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: Waiting for cloud-init to finish...
Nov 27 07:47:10 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: [INFO] No userdata found. Starting 1 node clusters, by default.
Nov 27 07:47:10 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: [INFO] Using instance type: i3.2xlarge
Nov 27 07:47:10 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: [INFO] meta-data:instance-type: i3.2xlarge
Nov 27 07:47:10 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: [INFO] meta-data:local-ipv4: 172.16.99.162
Nov 27 07:47:10 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: [INFO] meta-data:public-hostname: ec2-54-87-142-161.compute-1.amazonaws.com
Nov 27 07:47:10 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: [INFO] meta-data:ami-launch-index: 0
Nov 27 07:47:10 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: [INFO] meta-data:reservation-id: r-0852aca4b22f1a361
Nov 27 07:47:10 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: [INFO] Using cluster name: Test Cluster - No AMI Parameters
Nov 27 07:47:10 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: [INFO] Using cluster size: 1
Nov 27 07:47:10 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: [INFO] Using seed indexes: [0, 1, 1]
Nov 27 07:47:10 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: [INFO] Reflector loop...
Nov 27 07:47:10 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: [INFO] 11/27/19-07:47:10 Reflector: Received 1 of 1 responses from: [u'172.16.99.162']
Nov 27 07:47:10 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: [INFO] scylla.yaml configured.
Nov 27 07:47:12 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: meta-data=/dev/nvme0n1           isize=512    agcount=4, agsize=115966797 blks
Nov 27 07:47:12 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: =                       sectsz=512   attr=2, projid32bit=1
Nov 27 07:47:12 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: =                       crc=1        finobt=0, sparse=0
Nov 27 07:47:12 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: data     =                       bsize=4096   blocks=463867187, imaxpct=5
Nov 27 07:47:12 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: =                       sunit=0      swidth=0 blks
Nov 27 07:47:12 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: naming   =version 2              bsize=4096   ascii-ci=0 ftype=1
Nov 27 07:47:12 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: log      =internal log           bsize=4096   blocks=226497, version=2
Nov 27 07:47:12 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: =                       sectsz=512   sunit=0 blks, lazy-count=1
Nov 27 07:47:12 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: realtime =none                   extsz=4096   blocks=0, rtextents=0
Nov 27 07:47:12 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: Creating XFS volume for scylla using 1 disk(s): /dev/nvme0n1
Nov 27 07:47:12 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: [INFO] RAID Array containing ['nvme0n1'] not found. Creating...
Nov 27 07:47:12 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: Traceback (most recent call last):
Nov 27 07:47:12 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: File "/opt/scylladb/scripts/libexec/scylla_sysconfig_setup", line 76, in <module>
Nov 27 07:47:12 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: rps_cpus = out('{} --tune net --nic {} --get-cpu-mask'.format(perftune_base_command(), i
Nov 27 07:47:12 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: File "/opt/scylladb/scripts/scylla_util.py", line 495, in perftune_base_command
Nov 27 07:47:12 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: disk_tune_param = "--tune disks " + " ".join("--dir {}".format(d) for d in get_scylla_di
Nov 27 07:47:12 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: File "/opt/scylladb/scripts/scylla_util.py", line 478, in get_scylla_dirs
Nov 27 07:47:12 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: raise Exception("{}: at least one directory has to be set in 'data_file_directory'".form
Nov 27 07:47:12 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: Exception: /etc/scylla/scylla.yaml: at least one directory has to be set in 'data_file_d
Nov 27 07:47:13 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: Traceback (most recent call last):
Nov 27 07:47:13 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: File "/opt/scylladb/scylla-ami/libexec/scylla_ami_setup", line 32, in <module>
Nov 27 07:47:13 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: run('/opt/scylladb/scripts/scylla_sysconfig_setup --nic eth0 --setup-nic --ami')
Nov 27 07:47:13 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: File "/opt/scylladb/scripts/scylla_util.py", line 334, in run
Nov 27 07:47:13 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: return subprocess.check_call(cmd, shell=shell, stdout=stdout, stderr=stderr, env=scylla_
Nov 27 07:47:13 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: File "/opt/scylladb/python3/lib64/python3.7/subprocess.py", line 363, in check_call
Nov 27 07:47:13 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: raise CalledProcessError(retcode, cmd)
Nov 27 07:47:13 ip-172-16-99-162.ec2.internal scylla_ami_setup[3126]: subprocess.CalledProcessError: Command '['/opt/scylladb/scripts/scylla_sysconfig_setup',
Nov 27 07:47:13 ip-172-16-99-162.ec2.internal systemd[1]: scylla-ami-setup.service: main process exited, code=exited, status=1/FAILURE
Nov 27 07:47:13 ip-172-16-99-162.ec2.internal systemd[1]: Failed to start Scylla AMI Setup.
Nov 27 07:47:13 ip-172-16-99-162.ec2.internal systemd[1]: Unit scylla-ami-setup.service entered failed state.
Nov 27 07:47:13 ip-172-16-99-162.ec2.internal systemd[1]: scylla-ami-setup.service failed.
```

@syuu1228 @hagitsegev @slivne

**Comment by amoskong** [comment]
This issue was introduced by https://github.com/scylladb/scylla/commit/e0f40ed16aa4efa5892dfa9025b0dc20adbd60c2

It commented data_file_directories & commitlog_directory, which already have the same default value in code. But  scylla_util.py: perftune_base_command() checks both data_file_directories & commitlog_directory, it will raise exception if data_file_directories or commitlog_directory is empty.

 ```
diff --git a/conf/scylla.yaml b/conf/scylla.yaml
index 70844fa94d..429f141155 100644
--- a/conf/scylla.yaml
+++ b/conf/scylla.yaml
@@ -25,15 +25,19 @@
....
 # Directory where Scylla should store data on disk.
-# If not set, the default directory is /var/lib/scylla/data.
-data_file_directories:
-    - /var/lib/scylla/data
+# data_file_directories:
+#    - /var/lib/scylla/data
 
 # commit log.  when running on magnetic HDD, this should be a
 # separate spindle than the data directories.
-# If not set, the default directory is /var/lib/scylla/commitlog.
-commitlog_directory: /var/lib/scylla/commitlog
+# commitlog_directory: /var/lib/scylla/commitlog

....

```

```
./dist/common/scripts/scylla_util.py

465 def get_scylla_dirs():
466     """
467     Returns a list of scylla directories configured in /etc/scylla/scylla.yaml.
468     Verifies that mandatory parameters are set.
469     """
470     scylla_yaml_name = '/etc/scylla/scylla.yaml'
471     y = yaml.safe_load(open(scylla_yaml_name))
472 
473     # Check that mandatory fields are set
474     if 'data_file_directories' not in y or \
475             not y['data_file_directories'] or \
476             not len(y['data_file_directories']) or \
477             not " ".join(y['data_file_directories']).strip():
478         raise Exception("{}: at least one directory has to be set in 'data_file_directory'".format(scylla_yaml_name))
479     if 'commitlog_directory' not in y or not y['commitlog_directory']:
480         raise Exception("{}: 'commitlog_directory' has to be set".format(scylla_yaml_name))
...
```

**Comment by xemul** [comment]
These are now both optional. I can fix the script, do you want me to send a patch or make a pull request?

**Comment by amoskong** [comment]
> These are now both optional. I can fix the script, do you want me to send a patch or make a pull request?

Can you help to review https://github.com/scylladb/scylla/pull/5358 ？

**Comment by amoskong** [comment]
Fixed by https://github.com/scylladb/scylla/pull/5358

**Comment by slivne** [comment]
fixed by d4a26f2ad08891d57b9a4ac41de9eac8bf1a5fcb
